### PR TITLE
Update outdated preferences on startup

### PR DESCRIPTION
**Pull request by fxha**
| Q                | A
| ---------------- | ---
| Bug fix?         | yes
| New feature?     | yes
| Fixed tickets    | #100
| License          | MIT

### CODE REVIEW

File: src/main/preference.js [hunk 1]
```diff
@@ -57,6 +58,29 @@ class Preference {
...
     })
   }
+
+  update () {
+    if (this.requiresUpdate) {
+      this.requiresUpdate = false
+
+      // write latest settings to filesystem
+      const userSettings = this.getAll()
+      return this.setItem('theme', userSettings.theme)
```
Comment (Jocs) [inline src/main/preference.js]: A little confused, only set the theme?

File: src/main/actions/marktext.js [hunk 2]
```diff
@@ -47,7 +47,7 @@ autoUpdater.on('update-downloaded', () => {
 })
 
 export const userSetting = (menuItem, browserWindow) => {
-  createWindow(userPreference.userDataPath)
+  userPreference.update().then(() => createWindow(userPreference.userDataPath))
```
Comment (Jocs) [inline src/main/actions/marktext.js]: I think the best place to update user preference is the `Preference ` constructor.

File: src/main/preference.js [hunk 3]
```diff
@@ -22,19 +23,19 @@ class Preference {
   }
```
Comment (Jocs) [inline src/main/preference.js]: Maybe a `checkNeedUpdate` method is needed.

File: src/main/actions/marktext.js (hunk 2 above)
Comment (fxha) [reply src/main/actions/marktext.js]: Yup, but it that safe (side effects; "write" is an non-blocking system call)?

One solution would be to load setting into cache and update local file. 

```
ctor {
  ...

  getAll()
  if (requiresUpdate) {
    this.update()
  }
}
```

File: src/main/preference.js (hunk 1 above)
Comment (fxha) [reply src/main/preference.js]: Yes... `cache` has the latest settings combined with user settings and `this.setItem('theme', userSettings.theme)` triggers the save. It's not a perfect solution but no redundant code.

File: src/main/preference.js (hunk 3 above)
Comment (fxha) [reply src/main/preference.js]: What did you mean? A update is needed if `defaultSettings.keyNames != userSettings.keyNames`.

File: src/main/actions/marktext.js (hunk 2 above)
Comment (Jocs) [reply src/main/actions/marktext.js]: >side effects; "write" is an non-blocking system call

You mean `fs.write` ?

File: src/main/preference.js (hunk 3 above)
Comment (Jocs) [reply src/main/preference.js]: My opinion is that first check if it is needed to update in Preference constructor.

If it is needed, then update the preferences.md in `userDataPath`.

File: src/main/actions/marktext.js (hunk 2 above)
Comment (fxha) [reply src/main/actions/marktext.js]: Yes, in general system write.

File: src/main/preference.js (hunk 3 above)
Comment (fxha) [reply src/main/preference.js]: Yes, this is done by `this.requiresUpdate = !hasSameKeys(defaultSettings, userSettings)`.

**Comment by fxha** [comment]
@Jocs I updated this PR.

File: src/main/preference.js [hunk 4]
```diff
@@ -57,6 +65,28 @@ class Preference {
       })
     })
   }
+
+  update () {
```
Comment (Jocs) [inline src/main/preference.js]: Another question, I still think you should delete the outdated user setting and add the new use setting to `preferences.md` in `userDataPath `.

eg,

this old user preferences is:
```json
color: 'red',
theme: 'dark'
```

and the new user preferences is:

```json
darkColor: '#ffffff',
lightColor: '#000000',
theme: 'light'
```

The `color` in the setting is out of date, so we should delete it, and add `darkColor` and `lightColor` to `preferences.md`.

is it right ?

Comment (fxha) [reply src/main/preference.js]: I didn't know that we have outdated entries. Ok... let's say we create a array with a pair of `oldKeyName` and a function like this:

```
function (oldKey, oldValue) {
  // update key
  return newKey
}
```

Then we can replace old keys and/or convert them.

Comment (Jocs) [reply src/main/preference.js]: My fault, my before comments are not clear.

**Comment by fxha** [comment]
Updated change log. Can you please review?

**Comment by Jocs** [comment]
@fxha Sorry, These days, I am busy with my work project, I will review it tonight now.

**Comment by fxha** [comment]
@Jocs There's no hurry, so review it when you have time.

File: src/main/preference.js [hunk 5]
```diff
@@ -13,28 +13,51 @@ class Preference {
     this.cache = null
     this.staticPath = staticPath
     this.userDataPath = userDataPath
+    this.requiresUpdate = false
+
+    this.outdatedKeys = [
```
Comment (Jocs) [inline src/main/preference.js]: I maybe make you confused.

I gave the example of `color`, `darkColor` and `lightColor`.

It's just an example, and we should compare the old preferences and new preferences in `userDataPath` and `staticPath`, and compute the outdate preferences and newly added preferences.

Then we delete the outdate keys and add the new ones.

Comment (fxha) [reply src/main/preference.js]: Ok, outdated keys are now removed.

**Comment by Jocs** [comment]
@fxha FYI, merged and changed a little.

**Pull request merged**